pkg/apis/aws/helper/errors.go

```diff
@@ -0,0 +1,101 @@
+	retryableConfigurationProblemRegexp = regexp.MustCompile(`(?i)(is misconfigured and requires zero voluntary evictions|SDK.CanNotResolveEndpoint|The requested configuration is currently not supported)`)
+)
+
+// DetermineError determines the Gardener error codes for the given error and returns an ErrorWithCodes with the error and codes.
+func DetermineError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	// try to re-use codes from error
+	var coder gardencorev1beta1helper.Coder
+	if errors.As(err, &coder) {
+		return err
+	}
+
+	codes := DetermineErrorCodes(err)
+	if len(codes) == 0 {
+		return err
+	}
+
+	return gardencorev1beta1helper.NewErrorWithCodes(err, codes...)
+}
+
+// DetermineErrorCodes determines error codes based on the given error.
+func DetermineErrorCodes(err error) []gardencorev1beta1.ErrorCode {
+	var (
+		coder   gardencorev1beta1helper.Coder
+		message = err.Error()
+		codes   = sets.NewString()
+
+		knownCodes = map[gardencorev1beta1.ErrorCode]func(string) bool{
+			gardencorev1beta1.ErrorInfraUnauthenticated:          unauthenticatedRegexp.MatchString,
+			gardencorev1beta1.ErrorInfraUnauthorized:             unauthorizedRegexp.MatchString,
+			gardencorev1beta1.ErrorInfraQuotaExceeded:            quotaExceededRegexp.MatchString,
+			gardencorev1beta1.ErrorInfraRateLimitsExceeded:       rateLimitsExceededRegexp.MatchString,
+			gardencorev1beta1.ErrorInfraDependencies:             dependenciesRegexp.MatchString,
+			gardencorev1beta1.ErrorRetryableInfraDependencies:    retryableDependenciesRegexp.MatchString,
+			gardencorev1beta1.ErrorInfraResourcesDepleted:        resourcesDepletedRegexp.MatchString,
+			gardencorev1beta1.ErrorConfigurationProblem:          configurationProblemRegexp.MatchString,
+			gardencorev1beta1.ErrorRetryableConfigurationProblem: retryableConfigurationProblemRegexp.MatchString,
+		}
+	)
+
+	// try to re-use codes from error
+	if errors.As(err, &coder) {
+		for _, code := range coder.Codes() {
+			codes.Insert(string(code))
+			// found codes don't need to be checked any more
+			delete(knownCodes, code)
+		}
+	}
+
+	// determine error codes
+	for code, matchFn := range knownCodes {
+		if !codes.Has(string(code)) && matchFn(message) {
+			codes.Insert(string(code))
+		}
+	}
+
+	// compute error code list based on code string set
+	var out []gardencorev1beta1.ErrorCode
+	for _, c := range codes.List() {
+		out = append(out, gardencorev1beta1.ErrorCode(c))
+	}
+	return out
+}
```

Comment:
Why do we still have these functions in the light of https://github.com/gardener/gardener/pull/6912?

Reply:
It is draft I have not pushed recent changes, I was meanwhile looking into something else. I will push the change with the addressing review from the last PR so overall review effort is less.

Reply:
What is the value of such draft PR when it is not adapted at all? Typically, draft PRs can be used to get early feedback, but in this case this was not even your intention. Anyways ...